### PR TITLE
Remove _u_old from ScalarKernel/AuxScalarKernel and const-ify scalar vars

### LINKED ISSUE #17379: Do not request old values in ScalarKernel/AuxScalarKernel by default
## Reason
`ScalarKernel` and `AuxScalarKernel` obtain the older solution by default, which could lead to extraneous vector allocation that isn't really needed (think steady simulations with ScalarKernels).

## Design
Remove `_u_old` from `ScalarKernel`/`AuxScalarKernel` and add an old solution getter similarly to what was done in `AuxKernel`.

## Impact
Decreases memory allocation.

### LINKED ISSUE #17451: Const all the things in MooseVariableScalar and ScalarCoupleable
## Reason
Most methods in MooseVariableScalar and ScalarCoupleable are incredibly non-const correct.

## Design
Const all the things that we can.

## Impact
More const-correct code = better code.

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #17379 and closes #17451.

### CODE REVIEW

**Comment by loganharbour** [comment]
I'll create a patch for SAM tomorrow.

**Comment by permcody** [comment]
So what does it mean to be in "draft" and labeled as "ready for review"?

**Comment by loganharbour** [comment]
In this context, it means it's ready for review (MOOSE changes are done) but shouldn't be merged cause it'll break SAM.

**Comment by GiudGiud** [comment]
Is the new uOld() used somewhere in the framework? Or is this something for the apps?

Either way needs a test

**Comment by loganharbour** [comment]
> Is the new uOld() used somewhere in the framework? Or is this something for the apps?

I actually got around this by using it in the explicit case

File: framework/include/kernels/ScalarKernel.h [hunk 1]
```diff
@@ -39,6 +39,13 @@ class ScalarKernel : public ResidualObject, public ScalarCoupleable
    */
   virtual bool isActive() { return true; }
 
+  /**
+   * Retrieves the old value of the variable that this ScalarKernel operates on.
+   *
+   * Store this as a _reference_ in the constructor.
+   */
+  const VariableValue & uOld() const;
+
```
Comment (rwcarlsen) [inline framework/include/kernels/ScalarKernel.h:48]: Do you need this function to unbreak some particular app - or are you just assuming we'll eventually need it?

Comment (loganharbour) [reply framework/include/kernels/ScalarKernel.h:48]: Yeah - a few applications use the old solution.

**Review by rwcarlsen — DISMISSED** [review]
LGTM.  But I'm sure @GiudGiud can find something for you to tweak ;-). Looks like there are some modules tests to sort out though.

File: framework/include/variables/MooseVariableScalar.h [hunk 2]
```diff
@@ -157,11 +92,11 @@ class MooseVariableScalar : public MooseVariableBase
   /// Tagged vectors
   std::vector<VariableValue> _vector_tag_u;
   /// Only cache data when need it
-  std::vector<bool> _need_vector_tag_u;
+  mutable std::vector<bool> _need_vector_tag_u;
```
[the remark this answers is not in the document]
Comment (loganharbour) [reply framework/include/variables/MooseVariableScalar.h:95]: We do this everywhere already in Coupleable. We’re returning a const reference, so we should have the design be const. It’s private anyway

**Comment by loganharbour** [comment]
> @loganharbour - is this ready other than a few PRs to apps? Do you have PRs to failing apps ready to go?

There’s a few more. This turned out being a little more than I expected. Will ping when it’s ready.

File: framework/src/interfaces/ScalarCoupleable.C [hunk 3]
```diff
@@ -316,10 +318,28 @@ ScalarCoupleable::checkVar(const std::string & var_name)
...
+    if (comp < entry.size())
+      return entry[comp];
+    else
+      mooseError(_sc_name, ": Trying to get a non-existent component of variable '", var_name, "'");
+  }
+  else
+    mooseError(_sc_name, ": Trying to get a non-existent variable '", var_name, "'");
+}
+
+const MooseVariableScalar *
```
[the remark this answers is not in the document]
Comment (loganharbour) [reply framework/src/interfaces/ScalarCoupleable.C:334]: Looks like not. removed

**Comment by loganharbour** [comment]
This will be good to go once @andrsd does his weekly updates due to THM. I'm not in a rush, so Tuesday is fine.

**Comment by andrsd** [comment]
Apps should be updated with the THM fix, so this should be ready unless you are waiting on more things to happen.

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**